src/types.jl

```diff
@@ -75,12 +90,16 @@ end
 (::Type{D})(::Type{R}; kws...) where {R,D<:AbstractDimensions} = constructor_of(D){R}((tryrationalize(R, get(kws, k, zero(R))) for k in static_fieldnames(D))...)
 (::Type{D})(; kws...) where {R,D<:AbstractDimensions{R}} = constructor_of(D)(R; kws...)
 (::Type{D})(; kws...) where {D<:AbstractDimensions} = D(DEFAULT_DIM_BASE_TYPE; kws...)
-(::Type{D})(d::AbstractDimensions) where {R,D<:AbstractDimensions{R}} = D((getproperty(d, k) for k in static_fieldnames(D))...)
+function (::Type{D})(d::D2) where {R,D<:AbstractDimensions{R},D2<:AbstractDimensions}
+    issetequal(static_fieldnames(D), static_fieldnames(D2)) ||
+        error("Cannot create a dimensions of `$(D)` from `$(D2)`. Please write a custom method for construction.")
```

Comment:
I noticed this potential bug while writing this PR. It's unrelated to the goal of this PR but it was needed to fix some of the issues that came up, so I'm including it.